File: haskell/qu.hs
-- -*- coding: utf-8 -*-

-- | QueensモジュールはN Queensパズルの解を求めるための関数を提供します。 
module Queens (queens) where

import Data.List (permutations)


-- | queens関数はN Queensパズルの解を求めます。

-- queens関数の型定義。整数を受け取り、整数のリストのリストを返します。
-- Haskellは静的型付けですが、コンパイラが自動的に型を推測できるので、
-- 型定義は省略可能です。
queens :: Int -> [[Int]]

-- queens関数の本体
queens n = filter (safe . zip [1..]) $ permutations [1..n]

-- 上の式の (safe . zip ...) は「関数合成」といい、ラムダ式よりも簡潔に記述できます。


-- --------------------------------------------------------------------
-- プライベート関数（moduleで指定しなかった関数。モジュールの外からは呼べません）
-- --------------------------------------------------------------------

-- | safe関数は全ての駒が互いに相手を攻撃できないことを検査します。
--   攻撃できないときは安全という意味のTrueを、攻撃できる時はFalseを返します。
safe :: [(Int, Int)] -> Bool
safe []     = True
safe (x:xs) = all (safe2 x) xs && safe xs
  -- where句でローカル関数を定義できます。
  where safe2 (r1,c1) (r2,c2) = r1 + c1 /= r2 + c2 && r1 - c1 /= r2 - c2

File: erlang/qu.erl
%% -*- coding: utf-8 -*-

%%% @doc quモジュールはN Queensパズルの解を求めるための関数を提供します。

%% Erlangについてもっと知りたいときは、こちらを参照してください。
%%
%% 「Learn you some Erlang for great good!」 日本語翻訳版
%% http://ymotongpoo.appspot.com/lyse_ja/index.html


%%%
%%% コンパイラ向けの設定
%%%

-module(qu).
-compile([native, {hipe, ['O3']}]).  %% 高速化のためにネイティブコードにコンパイル

%% デバッグ用：本モジュールで定義した全ての関数が外部から参照できます。
-compile([export_all]).

%% 他のモジュールで定義されている関数をモジュール名抜きで参照します。
-import(qu_math,  [permutations/1]).
-import(qu_utils, [print_boards/2]).
-import(lists,    [all/2, filter/2, seq/2, zip/2]).


%%%
%%% 型定義
%%% Erlangは動的型付け言語ですが、-typeと-specを定義すると
%%% 静的コード解析ツールの「Dialyzer」で型情報が検証できるようになります。
%%%

-type cols() :: [integer()]. %% 全ての駒の位置を列で表します。


%%%
%%% 関数定義
%%% ここからがErlangプログラムです。
%%%

%% @doc queens関数はN Queensパズルの解を求めます。
%%      整数を引数に取り、解をcols型のリストで返します。
-spec queens(integer()) -> [cols()].
queens(N) ->
    L = seq(1, N),
    filter(fun(Cols) -> safe(zip(L, Cols)) end, permutations(L)).

%% @doc safe関数は駒が互いに相手を攻撃できないことを検査します。
%%      今は仮の実装として常にtrueを返すようにしています。
-spec safe(term()) -> true.
safe(_) ->
    true.

